File: widgets/spritebutton.lua
local _context = getContext()
local _painter = _context:getPainter()
local _style = _painter:getStyle()

exports('spriteButton', function(dict, name, text)
	_painter:setText(text)
	_painter:setTextOpts()

	local spriteButtonStyle = _style.spriteButton
	local sw = spriteButtonStyle.spriteWidth

	local w = _context:getWidgetWidth() or (_painter:calculateTextWidth() + _style.button.spacing * 2 + spriteButtonStyle.spacing + sw)
	local h = _style.widget.height

	_context:beginDraw(w, h)

	_painter:setColor(_context:isWidgetHovered() and _style.color.hover or _style.color.widget)
	_painter:drawRect(w, h)

	local sh = sw * GetAspectRatio()
	local so = (h - sh) / 2

	_painter:setColor(_style.color.primary)

	_painter:move(_style.button.spacing, so)
	_painter:drawSprite(dict, name, sw, sh)

	_painter:move(sw + spriteButtonStyle.spacing, -so)
	_painter:drawText()

	_context:endDraw()

	return _context:isWidgetClicked()
end)

File: widgets/spritebutton.js
exports('spriteButton', function (dict, name, text) {
	painter.setText(text)
	painter.setTextOpts()

	const spriteButtonStyle = style.spriteButton
	const sw = spriteButtonStyle.spriteWidth

	const w = context.getWidgetWidth() || (painter.calculateTextWidth() + style.button.spacing * 2 + spriteButtonStyle.spacing + sw)
	const h = style.widget.height

	context.beginDraw(w, h)

	painter.setColor(context.isWidgetHovered() ? style.color.hover : style.color.widget)
	painter.drawRect(w, h)

	const sh = sw * GetAspectRatio()
	const so = (h - sh) / 2

	painter.setColor(style.color.primary)

	painter.move(style.button.spacing, so)
	painter.drawSprite(dict, name, sw, sh)

	painter.move(sw + spriteButtonStyle.spacing, -so)
	painter.drawText()

	context.endDraw()

	return context.isWidgetClicked()
})
